drivers/hmis/app/models/hmis/form/form_processor.rb

```diff
@@ -427,4 +437,37 @@ def collect_record_validations(user: nil, household_members: nil)
 
     "#{container}.#{field}"
   end
+
+  # Ensure that a given field is valid for this FormDefinition
+  private def ensure_submittable_field!(container, field)
+    @mapped_form_fields ||= definition.link_id_item_hash.values.map(&:mapping).compact.
```

Comment:
Suggestions:
* mapped_form_fields could be a lookup table rather rather than an array. A lookup table (hash/set) should scale better with the number of fields in the form. Probably doesn't matter much until we have 100s of fields but might as well build it to support that.
* Maybe break the method up into two parts? Personally I like to wrap memoized variables in their own methods with the same name, that makes it harder to accidentally create an instance variable naming collision

Maybe something like this?

```ruby
  # @return <Hash{container_name=> Set<fields> }>
  private def mapped_form_fields
    @mapped_form_fields ||= {}.tap do |result|
      definition.link_id_item_hash.each_value do |item|
        mapping = item.mapping
        next unless mapping

        # convert the record_type to a "container name" that matches the form processor names
        container_name = if mapping.record_type
          enum = Types::Forms::Enums::RelatedRecordType.values[mapping.record_type]
          raise "Invalid record type '#{mapping.record_type}'" unless enum

          enum.description
        else
          owner_container_name
        end

        result[container_name] ||= Set.new
        # only one of these can be present, right?
        result[container_name].add(mapping.field_name || mapping.custom_field_key)
      end
    end
  end

  # Ensure that a given field is valid for this FormDefinition
  private def ensure_submittable_field!(container, field)
    # Find the FormItem Mapping that matches this field.
    # If it's not found, then this is not a valid submission.
    found_mapping = mapped_form_fields[container]&.include?(field)

    raise "Not a submittable field for Form Definition id #{definition.id} (#{container}.#{field})" unless found_mapping
  end
```

Reply:
thanks thats much cleaner